File: typ-trainning/Codeforces592/c.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long LL;
LL n, p, w, d;

void gcd(LL a, LL b, LL& d, LL& x, LL& y)  {
	if (!b) {d = a; x = 1; y = 0;}
	else { gcd(b, a%b, d, y, x); y -= x*(a/b);}  
}

__int128 Calc(__int128 x, LL y) {
	if (y == 1) return x;
	else return (x + y - 1) / y;
}

int main() {
	cin >> n >> p >> w >> d;
	LL g = __gcd(w, d);
	if (p % g) {
		puts("-1");
		return 0;
	}
	w /= g, d /= g, p /= g;
	LL tmp, y, x;
	gcd(w, d, tmp, x, y);
//	cout << x << " " << y << " " << p << endl;
	__int128 xx = x;
	__int128 yy = y;
	xx *= p, yy *= p;
	
	if (yy < 0) {
		__int128 add = Calc(-yy, w); 
		yy += (__int128) w * add;
		xx -= (__int128) d * add;
		x = xx, y = yy;
		if (xx < 0 || xx + yy > n) puts("-1");
		else cout << x << " " << y << " " << n - x - y << endl;
		return 0;
	}

	__int128 add = yy/w;
//	cout << add << endl;
	yy -= (__int128) w*add;
	xx += (__int128) d*add;
	x = xx, y = yy;
	if (xx < 0 || xx + yy > n) puts("-1");
	else cout << x << " " << y << " " << n - x - y << endl;
}

File: typ-trainning/Codeforces592/c.py
def gcd(a, b): 
	if b == 0:
		return a
	return gcd(b, a%b)

def exgcd(a, b, d):
	if b == 0:
		d = a
		x = 1
		y = 0
		return d, x, y
	d, y, x = exgcd(b, a%b, d)
	y = y - x*(a//b)
	return d, x, y

def Calc(x, y):
	if y == 1: 
		return x
	return (x + y - 1) // y

def main():
	n, p, w, d = map(int, input().split(" "))
	g = gcd(d, w)
	
#	print(g)
	if p % g > 0:
		print("-1");
		return
	p = p // g
	w = w // g
	d = d // g
	tmp = 0
	x = 0
	y = 0
	_d, x, y = exgcd(w, d, tmp)
	x = x * p
	y = y * p
#	print(str(x) + " " + str(y))
	if y < 0 :
		add = Calc(-y, w) 
#		print(add)
#		print(d)
		y += add * w
		x -= add * d
#		print(str(x) + " " + str(y))
		if x < 0 or x + y > n:
			print("-1")
		else :
			print(str(x) + " " + str(y) + " " + str(n - x - y))
		return
	
	add = y // w
	y -= add * w
	x += add * d
	if x < 0 or x + y > n :
		print("-1")
	else:
		print(str(x) + " " + str(y) + " " + str(n - x - y))
	return

main()
